pintc/src/asm_gen/asm_builder.rs

```diff
@@ -86,15 +90,40 @@ impl<'a> AsmBuilder<'a> {
         // Produce a map from input variables to this nodes to their indices in the concatinated
         // memory. Also compute the total amount of memory used so far.
         let mut total_memory_size = parents.iter().try_fold(0i64, |base, node| match node {
-            ComputeNode::NonLeaf { var, reads, .. } => {
+            ComputeNode::Var { var, reads, .. } => {
                 self.var_to_mem_idx.insert((var.name.clone(), *reads), base);
                 let size = var.expr.get_ty(contract).size(handler, contract)?;
                 Ok(base + 1 + size as i64)
             }
-            ComputeNode::Leaf { .. } => Ok(base),
+            ComputeNode::Constraint { .. } => Ok(base),
+            ComputeNode::Expr { expr, .. } => {
+                self.precomputed_expr_to_mem_idx.insert(*expr, base);
+                let size = expr.get_ty(contract).size(handler, contract)?;
+                Ok(base + 1 + size as i64)
+            }
         })?;
 
         let mut asm = Vec::new();
+
+        // First, collect all storage accesses used in the node expression, and allocate enough
+        // memory for all of them. We do this ahead of time so that we know exactly how the size of
+        // memory allocated. Due to short-circuting, this also means that some allocations may not
+        // be used, but this is okay for now.
+        for access in &expr.collect_storage_accesses(contract) {
+            let num_keys = access.get_ty(contract).storage_keys(handler, contract)? as i64;
+            let access_size = access.get_ty(contract).size(handler, contract)? as i64;
+            let alloc_size = 2 * num_keys // for each key, add the data address then its length
+                + access_size; // then lay out the data read from each key, sequentiallly
+
+            asm.extend([PUSH(alloc_size), ALOC, POP]);
```

Comment:
I think this ALOC might as well just be a single ALOC outside the loop, since we're throwing away the indices at runtime.  I noticed this recently when looking at map ASM gen.

Reply:
Fixed this.